**NAME : Gokul Nair**

**UID : 2018120039**

**DATE : 14-09-20**

**EXP : 5 B**



In [4]:
import pandas as pd
import numpy as np
import scipy as sp
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [5]:
! git clone https://github.com/defcom17/NSL_KDD


fatal: destination path 'NSL_KDD' already exists and is not an empty directory.


In [6]:
cd NSL_KDD

/content/NSL_KDD


In [7]:
Columns = ["duration","protocol_type","service","flag","src_bytes",
"dst_bytes","land","wrong_fragment","urgent","hot","num_failed_logins",
"logged_in","num_compromised","root_shell","su_attempted","num_root",
"num_file_creations","num_shells","num_access_files","num_outbound_cmds",
"is_host_login","is_guest_login","count","srv_count","serror_rate",
"srv_serror_rate","rerror_rate","srv_rerror_rate","same_srv_rate",
"diff_srv_rate","srv_diff_host_rate","dst_host_count","dst_host_srv_count",
"dst_host_same_srv_rate","dst_host_diff_srv_rate","dst_host_same_src_port_rate",
"dst_host_srv_diff_host_rate","dst_host_serror_rate","dst_host_srv_serror_rate",
"dst_host_rerror_rate","dst_host_srv_rerror_rate","label","difficulty"]

In [8]:
Data = pd.read_csv("KDDTrain+.csv", header=None, names = Columns)

In [9]:
Data.shape
Data.columns

Index(['duration', 'protocol_type', 'service', 'flag', 'src_bytes',
       'dst_bytes', 'land', 'wrong_fragment', 'urgent', 'hot',
       'num_failed_logins', 'logged_in', 'num_compromised', 'root_shell',
       'su_attempted', 'num_root', 'num_file_creations', 'num_shells',
       'num_access_files', 'num_outbound_cmds', 'is_host_login',
       'is_guest_login', 'count', 'srv_count', 'serror_rate',
       'srv_serror_rate', 'rerror_rate', 'srv_rerror_rate', 'same_srv_rate',
       'diff_srv_rate', 'srv_diff_host_rate', 'dst_host_count',
       'dst_host_srv_count', 'dst_host_same_srv_rate',
       'dst_host_diff_srv_rate', 'dst_host_same_src_port_rate',
       'dst_host_srv_diff_host_rate', 'dst_host_serror_rate',
       'dst_host_srv_serror_rate', 'dst_host_rerror_rate',
       'dst_host_srv_rerror_rate', 'label', 'difficulty'],
      dtype='object')

In [10]:
Data.describe

<bound method NDFrame.describe of         duration protocol_type  ...    label difficulty
0              0           tcp  ...   normal         20
1              0           udp  ...   normal         15
2              0           tcp  ...  neptune         19
3              0           tcp  ...   normal         21
4              0           tcp  ...   normal         21
...          ...           ...  ...      ...        ...
125968         0           tcp  ...  neptune         20
125969         8           udp  ...   normal         21
125970         0           tcp  ...   normal         18
125971         0           tcp  ...  neptune         20
125972         0           tcp  ...   normal         21

[125973 rows x 43 columns]>

In [11]:
from pandas_profiling import  ProfileReport


In [12]:
!pip install pandas-profiling==2.7.1

In [13]:
Data = Data.reset_index(level=0, drop=True).reset_index()

In [14]:
from sklearn import preprocessing
Data.protocol_type =preprocessing.LabelEncoder().fit_transform(Data["protocol_type"])
Data.service = preprocessing.LabelEncoder().fit_transform(Data["service"])
Data.flag = preprocessing.LabelEncoder().fit_transform(Data["flag"])
Data.label = preprocessing.LabelEncoder().fit_transform(Data["label"])

In [15]:
X = Data[Columns].to_numpy()
y = Data.label.to_numpy()

In [16]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test=train_test_split(X,y,test_size=0.2)

In [17]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix,classification_report
from sklearn.preprocessing import LabelEncoder
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import LinearSVC
from sklearn.neighbors import KNeighborsClassifier
%matplotlib inline

In [18]:
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier(max_depth=2, random_state=0)
clf.fit(X, y)
Score = clf.score(X_test,y_test)
print(Score*100)

86.38221869418535


In [19]:
#classifier
from sklearn.linear_model import LogisticRegression
clf=LogisticRegression()
import time
st=time.time()
#train the model
clf.fit(X_train,y_train)
et=time.time()
print("Training Time:", et-st)

Training Time: 56.37511205673218


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [20]:
#score
#Prediction of values using the model
#Model Prediction
st=time.time()
ypred=clf.predict(X_test)
et=time.time()
print ("Prediction Time:", et - st)
clf.score(X_test,y_test)

Prediction Time: 0.015130043029785156


0.8322683072038103

In [21]:
#model accuracy
acc_clf = 100 * clf.score(X_test,y_test)
print("Model Score : {:.3f} %".format(acc_clf))

Model Score : 83.227 %


In [22]:
#confusion matrix
from sklearn.metrics import confusion_matrix, classification_report
confusion_matrix(ypred,y_test)

array([[    0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0],
       [    0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0],
       [    0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0],
       [    0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0],
       [    0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0],
       [    0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0

In [23]:
print(classification_report(ypred,y_test))


              precision    recall  f1-score   support

           0       0.00      0.00      0.00         0
           1       0.00      0.00      0.00         0
           2       0.00      0.00      0.00         0
           3       0.00      0.00      0.00         0
           4       0.00      0.00      0.00         0
           5       0.00      0.00      0.00         0
           6       0.00      0.00      0.00         0
           7       0.00      0.00      0.00         0
           8       0.00      0.00      0.00         0
           9       0.93      0.89      0.91      8697
          10       0.00      0.00      0.00         0
          11       0.97      0.82      0.89     15851
          12       0.00      0.00      0.00         0
          13       0.00      0.00      0.00         0
          14       0.00      0.00      0.00         0
          15       0.34      0.30      0.32       647
          16       0.00      0.00      0.00         0
          17       0.00    

/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [24]:
#Creating the model : Decision Tree Classifier
dtc = DecisionTreeClassifier()
st = time.time()
#Fitting the data into the model
dtc.fit(X_train,y_train)
et = time.time()
print("Training time : {:.3f} s".format(et-st))

Training time : 0.291 s


In [25]:
#Prediction of values using the model
st = time.time()
y_pred = dtc.predict(X_test)
et = time.time()
print("Prediction time : {:.3f} s".format(et-st))
print("Predicted Values : ")
y_pred = pd.DataFrame(y_pred, columns = ["value"])
y_pred.groupby('value')['value'].count()

Prediction time : 0.009 s
Predicted Values : 


value
0       186
1         3
2         4
3        10
4         1
5       709
6         5
7         2
8         3
9      8252
10      297
11    13513
12        1
13        1
14       39
15      579
16        2
17      709
18      531
20      165
21      179
22        4
Name: value, dtype: int64

In [26]:
#model accuracy
acc_dtc = 100 * dtc.score(X_test,y_test)
print("Model Score : {:.3f} %".format(acc_dtc))

Model Score : 99.992 %


In [27]:
#Confusion Matrix
confusion_matrix(y_pred,y_test)

array([[  186,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0],
       [    0,     3,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0],
       [    0,     0,     3,     0,     1,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0],
       [    0,     0,     0,    10,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0],
       [    0,     0,     0,     1,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0],
       [    0,     0,     0,     0,     0,   709,     0,     0,     0,
            0,     0,     0,     0

In [28]:
#Classification Report
print(classification_report(y_pred,y_test))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00       186
           1       1.00      1.00      1.00         3
           2       1.00      0.75      0.86         4
           3       0.91      1.00      0.95        10
           4       0.00      0.00      0.00         1
           5       1.00      1.00      1.00       709
           6       1.00      1.00      1.00         5
           7       1.00      1.00      1.00         2
           8       1.00      1.00      1.00         3
           9       1.00      1.00      1.00      8252
          10       1.00      1.00      1.00       297
          11       1.00      1.00      1.00     13513
          12       1.00      1.00      1.00         1
          13       1.00      1.00      1.00         1
          14       1.00      1.00      1.00        39
          15       1.00      1.00      1.00       579
          16       1.00      1.00      1.00         2
          17       1.00    

In [29]:
#Creating the model : Random Forest Classifier
rfc = RandomForestClassifier()
st = time.time()
#Fitting the data into the model
rfc.fit(X_train,y_train)
et = time.time()
print("Training time : {:.3f} s".format(et-st))

Training time : 7.438 s


In [30]:
#Prediction of values using the model
st = time.time()
y_pred = rfc.predict(X_test)
et = time.time()
print("Prediction time : {:.3f} s".format(et-st))
print("Predicted Values : ")
y_pred = pd.DataFrame(y_pred, columns = ["value"])
y_pred.groupby('value')['value'].count()

Prediction time : 0.484 s
Predicted Values : 


value
0       186
1         5
2         2
3        10
5       709
6         5
8         3
9      8252
10      297
11    13516
12        2
13        1
14       39
15      579
16        1
17      709
18      531
20      165
21      179
22        4
Name: value, dtype: int64

In [31]:
#model accuracy
acc_rfc = 100 * rfc.score(X_test,y_test)
print("Model Score : {:.3f} %".format(acc_rfc))

Model Score : 99.972 %


In [32]:
#Confusion Matrix
confusion_matrix(y_pred,y_test)

array([[  186,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0],
       [    0,     3,     0,     0,     0,     0,     0,     1,     1,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0],
       [    0,     0,     2,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0],
       [    0,     0,     0,    10,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0],
       [    0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0],
       [    0,     0,     0,     0,     0,   709,     0,     0,     0,
            0,     0,     0,     0

In [33]:
#Classification Report
print(classification_report(y_pred,y_test))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00       186
           1       1.00      0.60      0.75         5
           2       0.67      1.00      0.80         2
           3       0.91      1.00      0.95        10
           4       0.00      0.00      0.00         0
           5       1.00      1.00      1.00       709
           6       1.00      1.00      1.00         5
           7       0.00      0.00      0.00         0
           8       0.67      0.67      0.67         3
           9       1.00      1.00      1.00      8252
          10       1.00      1.00      1.00       297
          11       1.00      1.00      1.00     13516
          12       1.00      0.50      0.67         2
          13       1.00      1.00      1.00         1
          14       1.00      1.00      1.00        39
          15       1.00      1.00      1.00       579
          16       0.50      1.00      0.67         1
          17       1.00    

/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [34]:
#Creating model : KNN Classifier
clf = KNeighborsClassifier()
st = time.time()
#Fitting the data into the model
clf.fit(X_train,y_train)
et = time.time()
print("Training time : {:.3f} s".format(et-st))

Training time : 4.909 s


In [35]:
#Prediction of values using the model
st = time.time()
y_pred = clf.predict(X_test)
et = time.time()
print("Predicted Values : ")
y_pred = pd.DataFrame(y_pred, columns = ["value"])
y_pred.groupby('value')['value'].count()

Predicted Values : 


value
0       183
3        11
5       721
6         5
7         2
9      8287
10      293
11    13522
13        1
14       40
15      532
17      714
18      532
20      167
21      182
22        3
Name: value, dtype: int64

In [36]:
#Prediction Time
print("Prediction time : {:.3f} s".format(et-st))

Prediction time : 8.415 s


In [37]:
#model accuracy
acc_knn = 100 * clf.score(X_test,y_test)
print("Model Score : {:.3f} %".format(acc_knn))

Model Score : 99.250 %


In [38]:
#Confusion Matrix
confusion_matrix(y_pred,y_test)

array([[  182,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     1,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0],
       [    0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0],
       [    0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0],
       [    0,     0,     0,    11,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0],
       [    0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0],
       [    0,     0,     0,     0,     0,   698,     0,     0,     0,
            0,    16,     7,     0

In [39]:
#Classification Report
print(classification_report(y_pred,y_test))

              precision    recall  f1-score   support

           0       0.98      0.99      0.99       183
           1       0.00      0.00      0.00         0
           2       0.00      0.00      0.00         0
           3       1.00      1.00      1.00        11
           4       0.00      0.00      0.00         0
           5       0.98      0.97      0.98       721
           6       1.00      1.00      1.00         5
           7       0.00      0.00      0.00         2
           8       0.00      0.00      0.00         0
           9       1.00      1.00      1.00      8287
          10       0.95      0.96      0.95       293
          11       1.00      1.00      1.00     13522
          12       0.00      0.00      0.00         0
          13       1.00      1.00      1.00         1
          14       1.00      0.97      0.99        40
          15       0.89      0.97      0.93       532
          16       0.00      0.00      0.00         0
          17       0.96    

/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [40]:
#SVM
from sklearn import svm

In [41]:
from sklearn.svm import SVC

In [ ]:
svclassifier = SVC(kernel='linear')
svclassifier.fit(X_train, y_train)

In [ ]:
y_pred = svclassifier.predict(X_test)

In [ ]:
from sklearn.metrics import classification_report, confusion_matrix
print(confusion_matrix(y_test,y_pred))
print(classification_report(y_test,y_pred))

# **Result**
Accuracy of Random Forest Classifier - 99.972 %

Accuracy of Decision Tree Classifier- 99.992 %

Accuracy of Logistic Regression - 84.854 %

Accuracy of KNN -  99.325 %

**Conclusion:-**
 1. From the following experiment conducted above we can infer that machine learning can be used to analyze data sets pertaining to cyber security and know what kind of attack is happening based on many parameters executed above such as count, flag, protocol and many more and hence a model can be created to help against these cyber attacks and can greatly improve the syber security ifnfrastructure. 

 2. In the above experiment we have also learned about various classification models and the best suited model for this experiment is the Decision Tree 
Classifier and the least suited model being Logistic Regression Classifier.

